tsl/src/compression/compression.c

```diff
@@ -2239,118 +2236,559 @@ compressed_insert_key_columns(Relation relation)
 	return indexattrs;
 }
 
-void
-decompress_batches_for_insert(const ChunkInsertState *cis, TupleTableSlot *slot)
+/* This method is used to find matching index on compressed chunk
+ * and build scan keys from the slot data
+ */
+static ScanKeyData *
+build_index_scankeys_using_slot(Oid hypertable_relid, Relation in_rel, Relation out_rel,
+								CompressionSettings *settings, TupleTableSlot *slot,
+								Relation *result_index_rel, int *num_scan_keys)
 {
-	/*
-	 * This is supposed to be called with the actual tuple that is being
-	 * inserted, so it cannot be empty.
-	 */
-	Assert(!TTS_EMPTY(slot));
+	List *index_oids;
+	ListCell *lc;
+	ScanKeyData *scankeys = NULL;
+	/* get list of indexes defined on compressed chunk */
+	index_oids = RelationGetIndexList(in_rel);
+	*num_scan_keys = 0;
 
-	Relation out_rel = cis->rel;
+	foreach (lc, index_oids)
+	{
+		Relation index_rel = index_open(lfirst_oid(lc), AccessShareLock);
+		bool matches = false;
 
-	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+		/* Must have at least two attributes. */
+		if (index_rel->rd_index->indnatts < 2)
+		{
+			index_close(index_rel, AccessShareLock);
+			continue;
+		}
+
+		scankeys = palloc0((index_rel->rd_index->indnatts) * sizeof(ScanKeyData));
+
+		for (int i = 0; i < index_rel->rd_index->indnatts; i++)
+		{
+			AttrNumber attnum = AttrOffsetGetAttrNumber(i);
+			char *attname = get_attname(RelationGetRelid(index_rel), attnum, false);
+
+			/* If we are at the last attribute, check its the sequence number attribute.
+			 * This means we found all other attributes on the hypertable and this could be our
+			 * index.
+			 */
+			if (index_rel->rd_index->indnatts - 1 == i)
+			{
+				if (strcmp(attname, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME) == 0)
+					matches = true;
+				break;
+			}
```

Comment:
If some starting prefix of index columns are segmentby columns, we can use the index for lookups by these segmentbys, no matter what index columns follow, right? Maybe we can simplify/generalize this condition accordingly.

Reply:
hmm ideally we want the index with the most columns matching the unique constraints, bonus points for considering selectivity, and we could cache this selection for the chunk, but for the common case we only create 1 index on compressed chunk so maybe this is overkill

Reply:
I think the general approach is even simpler than what we have now: count the length of prefix which consists of segmentbys, then choose the index that has the most. Or we could count the selectivity indeed in the same way, that would be a nice addition given that we have proper statistics for the segmentby columns.

Reply:
This will be removed with my upcoming work for removing sequence numbers so I'll leave it for now.